src/PubSubClient.cpp

```diff
@@ -174,12 +186,16 @@ boolean PubSubClient::connect(const char *id, const char *user, const char *pass
             write(MQTTCONNECT,buffer,length-5);
 
             lastInActivity = lastOutActivity = millis();
-
+			
             while (!_client->available()) {
                 unsigned long t = millis();
                 if (t-lastInActivity >= ((int32_t) MQTT_SOCKET_TIMEOUT*1000UL)) {
                     _state = MQTT_CONNECTION_TIMEOUT;
                     _client->stop();
+
+                    // Next time, Try to Switch Protocol if client not responding 
+                    protocolType = (protocolType==MQTT_VERSION_3_1_1) ? MQTT_VERSION_3_1 : MQTT_VERSION_3_1_1; 
```

Comment:
Hi,
1st I apologize for not having discussed this earlier..
A you may realize I recently subscibed to github.. hence I'm not yet fully
confident about the way to go to proactively contribute to open-source..

Just few words why I need to implement this: I'm doing a domotic controller
for the blinds in my house.. and you could understand that it is hard to
mantain it once hide in the walls..
So for example I know that in few months I will upgrade my MQTT broker from
12.04 to 16.04 and thus I know that I should eventually recompile
everithing on MCU due to change in protocol..
So i'm trying to compile a FW that still has low memory footprint yet being
a little adaptable..

OK, now I think I will go for the WE..
Have a nice evening and sorry again for the misunderstanding..

Mirko.


On Fri, Jan 29, 2016 at 7:34 PM, knolleary <notifications@github.com> wrote:

> In src/PubSubClient.cpp
> <https://github.com/knolleary/pubsubclient/pull/123#discussion_r51296918>:
>
> >              while (!_client->available()) {
> >                  unsigned long t = millis();
> >                  if (t-lastInActivity >= ((int32_t) MQTT_SOCKET_TIMEOUT*1000UL)) {
> >                      _state = MQTT_CONNECTION_TIMEOUT;
> >                      _client->stop();
> > +
> > +                    // Next time, Try to Switch Protocol if client not responding
> > +                    protocolType = (protocolType==MQTT_VERSION_3_1_1) ? MQTT_VERSION_3_1 : MQTT_VERSION_3_1_1;
>
> Need to put some more thought on if automatically switching is good or
> not. Typically, you will know (or discover the first time you test) whether
> the protocol you are using is supported by the broker you are pointed at.
> You are highly unlikely to need to change protocols at runtime after the
> initial setup.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/knolleary/pubsubclient/pull/123/files#r51296918>.
>